compiler/src/codegen/emitmod.re

```diff
@@ -1,8 +1,72 @@
 open Grain_typed;
 open Grain_utils;
-open Compmod;
+open Mashtree;
 
-let emit_module = ({asm, signature}, outfile) => {
+type bad_object =
+  | InvalidMagic
+  | InvalidVersion
+  | Corrupted;
+
+exception BadObject(bad_object);
+
+let emit_object = (mashed: mash_program, outfile) => {
+  Grain_utils.Fs_access.ensure_parent_directory_exists(outfile);
+  if (Config.debug^) {
+    let mash_string =
+      Sexplib.Sexp.to_string_hum @@ Mashtree.sexp_of_mash_program(mashed);
+    let mashfile = Filepath.String.replace_extension(outfile, "mashtree");
+    let oc = open_out(mashfile);
+    output_string(oc, mash_string);
+    close_out(oc);
+  };
+
+  let oc = open_out_bin(outfile);
+
+  output_string(oc, Cmi_format.magic);
+  let version_length = String.length(Config.version);
+  output_binary_int(oc, version_length);
+  output_string(oc, Config.version);
+  let cmi = Marshal.to_bytes(mashed.signature, []);
```

Comment:
Taking a little hint from the wasm bytecode how would we feel about having this be `sectionID ...sectionContents sectionLength` instead of just storing the data randomly, this would allow us to add more data in here easily in the future without breaking things while still being easy and quick to parse.